### PR TITLE
CORE-12672. [USETUP] IsThereAValidBootSector(): nits for r74512.

### PR DESCRIPTION
**Pull request by SergeGautherie**
## Purpose

`IsThereAValidBootSector()`: nits for r74512.

JIRA issue: [CORE-12672](https://jira.reactos.org/browse/CORE-12672)

## Proposed changes
- `ULONG Instruction` is no longer needed.
- Document `RtlZeroMemory()` use.
- Do tests in (better) documented order.
- Do not misuse `&=`, when `&&` is meant.

### CODE REVIEW

**Comment by SergeGautherie** [comment]
Cc @HBelusca

**Comment by paulsapps** [comment]
Build has failed `../../base/setup/usetup/bootsup.c:695:5: error: suggest parentheses around comparison in operand of '&' [-Werror=parentheses]`

File: base/setup/usetup/bootsup.c [hunk 1]
```diff
@@ -686,12 +687,12 @@ IsThereAValidBootSector(PWSTR RootPath)
...
-    Instruction = (*(PULONG)BootSector) & 0x00FFFFFF;
-    IsValid = (Instruction != 0x00000000);
+    /* Check the bootsector signature. */
+    if (*(PUSHORT)(BootSector + 0x01FE) != 0xAA55)
+        goto Quit;
 
-    /* Check the bootsector signature */
-    IsValid &= (*(PUSHORT)(BootSector + 0x1fe) == 0xaa55);
+    /* Check the instruction; we use a ULONG to read three bytes. */
+    IsValid = (*(PULONG)BootSector & 0x00FFFFFF != 0x00000000);
```
Comment (binarymaster) [inline base/setup/usetup/bootsup.c]:
```c
IsValid = ((*(PULONG)BootSector & 0x00FFFFFF) != 0x00000000);
```

**Comment by ThFabba** [comment]
I see no advantage to this change. It's code churn that seemingly has no effect on what the code does.

**Comment by binarymaster** [comment]
Yep, code logic didn't changed. It only became non-linear.

File: base/setup/usetup/bootsup.c (hunk 1 above)
Comment (HBelusca) [inline base/setup/usetup/bootsup.c]: The "&=" should have been "IsValid = IsValid && (...)" from the beginning.

**Comment by HBelusca** [comment]
See also https://github.com/reactos/reactos/blob/7794ea6817a2f9f318385b42cd820fc300c6c53a/base/setup/lib/bootsup.c#L590

**Comment by learn-more** [comment]
So I guess this would conflict with the setup_improvements branch?

**Comment by HBelusca** [comment]
@learn-more : well yes, but since they are simple (act on 3-4 lines only) they can be ported by hand.

File: base/setup/usetup/bootsup.c (hunk 1 above)
Comment (SergeGautherie) [reply base/setup/usetup/bootsup.c]: Done.

**Comment by SergeGautherie** [comment]
> no effect

@ThFabba, isn't '&=' the wrong operator? Or did I not fix it correctly?

> didn't change

@binarymaster, code logic is amended, to do what comment explains, not the opposite (order).

> conflict

@learn-more, after a quick look, the conflict should be 1-2 blank lines only/at-most, so trivial to resolve.

> setup_improvements branch

@HBelusca, #59 says it is not meant to be merged "soon".
If it now is, then the current PR can wait until then (or be submitted/merged there instead).
Otherwise, I think simple fixes (!= new features) should go to master directly/"asap".

**Comment by ThFabba** [comment]
`&=` is not the most straightforward operation to do, but it is not wrong. The value will be 1 or 0, so `&` and `&&` will have the same effect. `&&` short-circuits the operation, but (a) this is an insignificant optimization here, (b) as the `BootSector` pointer is not volatile, the compiler is free to perform that optimization anyway.
So yes, no effect.

**Comment by SergeGautherie** [comment]
> yes, no effect

You are so right, thanks! I mistakenly had `|=` logic in mind. Oups.
Nonetheless, I still think that using bitwise operators for boolean checks is confusing.

--

Now, is any of these accepted on master, or branch?

**Comment by HBelusca** [comment]
Let's consider this fixed (in the setup_improvements branch). The fix will go to master when the branch is merged. I haven't committed directly on master because:
- the fix is not a blocker;
- I don't want yet again to mangle with merge conflicts.

**Comment by SergeGautherie** [comment]
Eventually committed as a7a11dd.

**Pull request closed without merging**